File: js/tests/languages.test.ts
/**
 * Tests for meta-notation with various programming languages
 */

import { test, assert } from 'test-anywhere';
import { parse } from '../src/parser.js';
import { serialize } from '../src/serializer.js';
import type { Block, Sequence, DelimiterType } from '../src/types.js';

// Helper function to check if a delimiter type exists anywhere in the parsed result
function hasDelimiterType(sequence: Sequence, type: DelimiterType): boolean {
  for (const block of sequence) {
    if (block.type === type) {
      return true;
    }
    if (Array.isArray(block.content)) {
      if (hasDelimiterType(block.content, type)) {
        return true;
      }
    }
  }
  return false;
}

// JavaScript
test('parse JavaScript code', () => {
  const code = 'const greet = (name) => { return `Hello, ${name}!`; };';
  const result = parse(code);
  assert.ok(result.length > 0);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'backtick'));
  // Round-trip test
  assert.equal(serialize(result), code);
});

// Python
test('parse Python code', () => {
  const code = 'def calculate(x, y): return {"sum": x + y, "list": [x, y]}';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Go
test('parse Go code', () => {
  const code = 'func main() { fmt.Println("Hello, World!") }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Rust
test('parse Rust code', () => {
  const code = 'fn main() { let x = vec![1, 2, 3]; println!("{:?}", x); }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// C++
test('parse C++ code', () => {
  const code = 'int main() { std::cout << "Hello" << std::endl; return 0; }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Java
test('parse Java code', () => {
  const code = 'public class Main { public static void main(String[] args) { System.out.println("Hello"); } }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// C#
test('parse C# code', () => {
  const code = 'public void Test() { var list = new List<int> {1, 2, 3}; Console.WriteLine("Done"); }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Ruby
test('parse Ruby code', () => {
  const code = 'def greet(name); puts "Hello, #{name}!"; end';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// PHP
test('parse PHP code', () => {
  const code = 'function test($x) { return ["key" => "value"]; }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Swift
test('parse Swift code', () => {
  const code = 'func greet(name: String) -> String { return "Hello, \\(name)!" }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Kotlin
test('parse Kotlin code', () => {
  const code = 'fun main() { val list = listOf(1, 2, 3); println("Hello") }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// TypeScript
test('parse TypeScript code', () => {
  const code = 'const add = (a: number, b: number): number => { return a + b; };';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.equal(serialize(result), code);
});

// Scala
test('parse Scala code', () => {
  const code = 'def add(x: Int, y: Int): Int = { x + y }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.equal(serialize(result), code);
});

// Perl
test('parse Perl code', () => {
  const code = 'sub greet { my ($name) = @_; print "Hello, $name!\\n"; }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Haskell
test('parse Haskell code', () => {
  const code = 'main = putStrLn "Hello, World!"';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Lisp/Scheme
test('parse Lisp code', () => {
  const code = '(define (factorial n) (if (= n 0) 1 (* n (factorial (- n 1)))))';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.equal(serialize(result), code);
});

// Clojure
test('parse Clojure code', () => {
  const code = '(defn greet [name] (str "Hello, " name "!"))';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Lua
test('parse Lua code', () => {
  const code = 'function greet(name) return "Hello, " .. name end';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Elixir
test('parse Elixir code', () => {
  const code = 'def greet(name), do: "Hello, #{name}!"';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// R
test('parse R code', () => {
  const code = 'greet <- function(name) { paste("Hello,", name) }';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// MATLAB
test('parse MATLAB code', () => {
  const code = 'function y = square(x); y = x .^ 2; end';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'paren'));
  assert.equal(serialize(result), code);
});

// SQL
test('parse SQL code', () => {
  const code = 'SELECT name, age FROM users WHERE status = "active" ORDER BY created_at;';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// JSON
test('parse JSON', () => {
  const code = '{"name": "John", "age": 30, "tags": ["developer", "designer"]}';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'curly'));
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// YAML-like (with brackets)
test('parse YAML with brackets', () => {
  const code = 'dependencies: ["react", "typescript"]';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'square'));
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Shell/Bash
test('parse Bash code', () => {
  const code = 'echo "Hello, ${USER}!" | grep "Hello"';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'doubleQuote'));
  assert.equal(serialize(result), code);
});

// Markdown code blocks
test('parse Markdown with code blocks', () => {
  const code = 'Here is code: `const x = [1, 2, 3];` in backticks.';
  const result = parse(code);
  assert.ok(hasDelimiterType(result, 'backtick'));
  assert.equal(serialize(result), code);
});

File: rust/tests/languages_test.rs
//! Tests for meta-notation with various programming languages

use meta_notation::{parse, serialize, DelimiterType};

// Helper function to check if a delimiter type exists anywhere in the parsed result
fn has_delimiter_type(blocks: &[meta_notation::Block], dtype: &DelimiterType) -> bool {
    blocks.iter().any(|b| b.has_delimiter_type(dtype))
}

#[test]
fn test_parse_javascript_code() {
    let code = "const greet = (name) => { return `Hello, ${name}!`; };";
    let result = parse(code);
    assert!(result.len() > 0);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::Backtick));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_python_code() {
    let code = r#"def calculate(x, y): return {"sum": x + y, "list": [x, y]}"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_go_code() {
    let code = r#"func main() { fmt.Println("Hello, World!") }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_rust_code() {
    let code = r#"fn main() { let x = vec![1, 2, 3]; println!("{:?}", x); }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_cpp_code() {
    let code = r#"int main() { std::cout << "Hello" << std::endl; return 0; }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_java_code() {
    let code = r#"public class Main { public static void main(String[] args) { System.out.println("Hello"); } }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_csharp_code() {
    let code =
        r#"public void Test() { var list = new List<int> {1, 2, 3}; Console.WriteLine("Done"); }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_ruby_code() {
    let code = r#"def greet(name); puts "Hello, #{name}!"; end"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_php_code() {
    let code = r#"function test($x) { return ["key" => "value"]; }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_swift_code() {
    let code = r#"func greet(name: String) -> String { return "Hello, \(name)!" }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_kotlin_code() {
    let code = r#"fun main() { val list = listOf(1, 2, 3); println("Hello") }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_typescript_code() {
    let code = "const add = (a: number, b: number): number => { return a + b; };";
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_scala_code() {
    let code = "def add(x: Int, y: Int): Int = { x + y }";
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_perl_code() {
    let code = r#"sub greet { my ($name) = @_; print "Hello, $name!\n"; }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_haskell_code() {
    let code = r#"main = putStrLn "Hello, World!""#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_lisp_code() {
    let code = "(define (factorial n) (if (= n 0) 1 (* n (factorial (- n 1)))))";
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_clojure_code() {
    let code = r#"(defn greet [name] (str "Hello, " name "!"))"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_lua_code() {
    let code = r#"function greet(name) return "Hello, " .. name end"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_elixir_code() {
    let code = r#"def greet(name), do: "Hello, #{name}!""#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_r_code() {
    let code = r#"greet <- function(name) { paste("Hello,", name) }"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_matlab_code() {
    let code = "function y = square(x); y = x .^ 2; end";
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Paren));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_sql_code() {
    let code = r#"SELECT name, age FROM users WHERE status = "active" ORDER BY created_at;"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_json() {
    let code = r#"{"name": "John", "age": 30, "tags": ["developer", "designer"]}"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Curly));
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_yaml_with_brackets() {
    let code = r#"dependencies: ["react", "typescript"]"#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Square));
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_bash_code() {
    let code = r#"echo "Hello, ${USER}!" | grep "Hello""#;
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::DoubleQuote));
    assert_eq!(serialize(&result), code);
}

#[test]
fn test_parse_markdown_with_code_blocks() {
    let code = "Here is code: `const x = [1, 2, 3];` in backticks.";
    let result = parse(code);
    assert!(has_delimiter_type(&result, &DelimiterType::Backtick));
    assert_eq!(serialize(&result), code);
}
